### PR TITLE
RSDK-6825 - relative inputs incorrectly place frames

### PR DESCRIPTION
**Pull request by nfranczak**
https://viam.atlassian.net/browse/RSDK-6825

### CODE REVIEW

File: services/motion/builtin/builtin_test.go [hunk 1]
```diff
@@ -509,93 +509,6 @@ func TestObstacleReplanningGlobe(t *testing.T) {
 	}
 }
 
-func TestObstacleReplanningSlam(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (nfranczak) [reply services/motion/builtin/builtin_test.go:512]: yea, not super satisfying.. T^T

I will say that what we test in `TestObstacleReplanningSlam` is also tested in `TestObstacleReplanningGlobe`.
Further, `TestObstacleReplanningSlam` isn't really a unit test, rather it is an integration test.. so I am ok with removing it.

[the remark this answers is not in the document]
Comment (raybjork) [reply services/motion/builtin/builtin_test.go:512]: I talked to Miko about this.  It feels like this test in particular is a little duplicative of another test we have exercising replanning for MoveOnGlobe.  Since it seems at the moment that the flake is not an error in business logic and rather a result of the difficulty of writing integration tests, I didnt think it was super necessary to spend a lot of time on this when we are already testing the components it tests elsewhere.

File: motionplan/motionPlanner.go [hunk 2]
```diff
@@ -483,12 +483,10 @@ func CheckPlan(
...
 
 	// create a list of segments to iterate through
-	var segments []*ik.Segment
+	segments := make([]*ik.Segment, 0, len(poses))
 	if relative {
-		segments = make([]*ik.Segment, 0, len(poses)+1)
-
 		// get the inputs we were partway through executing
-		checkFrameGoalInputs, err := sf.mapToSlice(plan.Trajectory()[0])
+		checkFrameGoalInputs, err := sf.mapToSlice(plan.Trajectory()[1])
```
Comment (raybjork) [inline motionplan/motionPlanner.go:489]: why is this changing?

File: motionplan/motionPlanner.go [hunk 3]
```diff
@@ -549,8 +537,15 @@ func CheckPlan(
 		}, nil
 	}
 
+	var iterStartVal int
```
Comment (raybjork) [inline motionplan/motionPlanner.go:540]: This bit seems pretty fishy to me.  Why is this necessary

File: services/motion/builtin/move_request.go [hunk 4]
```diff
@@ -137,14 +137,14 @@ func (mr *moveRequest) AnchorGeoPose() *spatialmath.GeoPose {
 
 // execute attempts to follow a given Plan starting from the index percribed by waypointIndex.
 // Note that waypointIndex is an atomic int that is incremented in this function after each waypoint has been successfully reached.
-func (mr *moveRequest) execute(ctx context.Context, plan motionplan.Plan, waypointIndex *atomic.Int32) (state.ExecuteResponse, error) {
```
Comment (raybjork) [inline services/motion/builtin/move_request.go:140]: Why is this changing?

File: motionplan/motionPlanner.go (hunk 2 above)
Comment (nfranczak) [reply motionplan/motionPlanner.go:489]: within `move_request.go/obstaclesIntersectPlan` the plan object passed into this function used to be `motionplan.RemainingPlan(plan, waypointIndex)`.
Now, to take advantage of the new `motionplan.Plan` object we do `motionplan.RemainingPlan(plan, waypointIndex-1)`.

Suppose we have constructed some plan with the following inputs `[ [0, 0, 0], [a, b, c], [d, e, f], [0, 0, 0] ]` and the following poses `[ startPose, middlePose, goalPose, goalPose ]`
Let our current input be `[a, b, x]` such that `x < c`. Hence, our wayPointIndex would be 1.

Therefore, `remainingPlan := motionplan.RemainingPlan(plan, waypointIndex-1)` gives us back the entirety of the plan. So the inputs we were part-way through would be `remainingPlan.Traj()[1]`.

The motivation for `waypointIndex-1` is that `remainingPlan.Path()[0]` is the `startPose` i.e. the `formerRunningPose`, which we can now gain without having to determine the `lastPose` and `pathPosition`

File: motionplan/motionPlanner.go (hunk 3 above)
Comment (nfranczak) [reply motionplan/motionPlanner.go:540]: This was introduced because we changed the length of the `plan` object which is received by `CheckPlan` as passed in through `obstaclesIntersectPlan`. 

For relative cases, I believe it makes sense to pass in a plan which includes _not only_ the remaining inputs + poses left to be executed, **but also** the most recent already executed input and reached pose. The benefit of doing this is that we can determine the value of the `formerRunningPose` without having to also calculate the values of our `pathPosition` and `lastPose`. 
Note: we care about the most recently reached pose since that is the value we must compose with `frame.Transform(currentInputs)` to place us in absolute coordinates.

Ultimately, this allows for us to rely less on what controls are telling us about the state of the robot and instead use the immutable plan object to inform us.

However! `iterStartVal` is introduced solely for `TestArmGantryCheckPlan` since this is the only code path which tests `CheckPlan` for an already absolute plan.

File: services/motion/builtin/move_request.go [hunk 5]
```diff
@@ -304,6 +291,22 @@ func (mr *moveRequest) obstaclesIntersectPlan(
...
+			if err != nil {
+				return state.ExecuteResponse{}, err
+			}
+
+			// Note: the value of wayPointIndex is subject to change between when this function is first entered
+			// versus when CheckPlan is actually called.
+			// We load the wayPointIndex value to ensure that all information is up to date.
+			waypointIndex := int(mr.waypointIndex.Load())
+
+			remainingPlan, err := motionplan.RemainingPlan(plan, waypointIndex-1)
```
Comment (raybjork) [inline services/motion/builtin/move_request.go:305]: will `waypointIndex` ever be 1?

File: services/motion/builtin/move_request.go (hunk 4 above)
Comment (nfranczak) [reply services/motion/builtin/move_request.go:140]: The motivation for this spawned from recognizing that `waypointIndex` should not be an argument for `obstaclesIntersectPlan` or `execute` due to timing issues.

Suppose we have constructed some plan with the following inputs `[ [0, 0, 0], [a, b, c], [d, e, f], [0, 0, 0] ]` and the following poses `[ startPose, middlePose, goalPose, goalPose ]`.

Suppose that when we enter `obstaclesIntersectPlan` our `wayPointIndex == 1`. Hence, our currentInputs are `[a, b, x]` such that `x < c`. We then query the vision service + camera for detections.. [ some amount of time passes ]

We now have gained our detections but suppose that our currentInputs are now `[d, e, y]` such that `y < f`.
Then, the wayPointIndex within `execute` would be updated to `2`. but the wayPointIndex known by `obstaclesIntersectPlan` would be 1. This would result in an incorrect interpolation of the plan.

File: motionplan/motionPlanner.go (hunk 2 above)
Comment (raybjork) [reply motionplan/motionPlanner.go:489]: It took some thinking through but I think this makes sense to me.  I think this makes sense to me.  I think the key insight is that we always are currently executing `Trajectory()[waypointIndex]` and the object being passed into this function includes the previous point as well.  I think some documentation explaining that the remainingPlan object actually includes the previous point the plan originates from would be valuable

File: services/motion/builtin/move_request.go (hunk 5 above)
Comment (nfranczak) [reply services/motion/builtin/move_request.go:305]: Yes.

Suppose we have constructed some plan with the following inputs `[ [0, 0, 0], [a, b, c], [d, e, f], [0, 0, 0] ]`.

`waypointIndex := int(mr.waypointIndex.Load())` can be `1` if we call `obstaclesIntersectPlan` before we have finished executing `[a, b, c]`.

File: motionplan/motionPlanner.go (hunk 3 above)
Comment (raybjork) [reply motionplan/motionPlanner.go:540]: I see now.  Would it be possible to change the way we express absolute plans to get rid of the need to do this case structure?

Comment (raybjork) [reply motionplan/motionPlanner.go:540]: We could have the invariant that each plan always contains the first index being where it starts at (which now that I write it I think is already true)

File: services/motion/builtin/move_request.go (hunk 5 above)
Comment (raybjork) [reply services/motion/builtin/move_request.go:305]: I'm sorry I meant to write 0.  XD.  I'm trying to check if we're going to have array index problems here

File: motionplan/motionPlanner.go (hunk 2 above)
Comment (nfranczak) [reply motionplan/motionPlanner.go:489]: > I think some documentation explaining that the remainingPlan object actually includes the previous point the plan originates from would be valuable

good call out! I will update with a comment (both in move_request/obstaclesIntersectPlan and here)

File: services/motion/builtin/move_request.go (hunk 5 above)
Comment (nfranczak) [reply services/motion/builtin/move_request.go:305]: ahh I gotchu, I believe this is impossible ™️ . This is my motivation for thinking so:
https://github.com/viamrobotics/rdk/blob/main/services/motion/builtin/move_request.go#L755

Comment (raybjork) [reply services/motion/builtin/move_request.go:305]: I am convinced.  Thanks!

File: services/motion/builtin/move_request.go (hunk 4 above)
Comment (nfranczak) [reply services/motion/builtin/move_request.go:140]: Yes, I agree that this fixing a latent bug.

File: motionplan/motionPlanner.go (hunk 3 above)
Comment (biotinker) [reply motionplan/motionPlanner.go:540]: This is making me wonder if we've set up `CheckPlan` poorly.

What we have right now is a public function with the undocumented requirement of "pass in only the plan left to execute, except going back in time one waypoint". Which is a weird way to capture state.

I am wondering if perhaps we want to instead pass in entire plans, and pass in the currently executing index, instead. Not necessarily in this PR.

Comment (raybjork) [reply motionplan/motionPlanner.go:540]: This would make it more apparent how this function is meant to be used I think. Coming at the cost of being a little more verbose.  Would be fine with this or else just documenting a little better

Comment (biotinker) [reply motionplan/motionPlanner.go:540]: I am happy to do this as part of my course correction work. Since we'll need to pass segments into this function for that to work properly, there will be a significant reworking of `CheckPlan` as part of that.

File: services/motion/builtin/move_request.go [hunk 6]
```diff
@@ -304,6 +291,22 @@ func (mr *moveRequest) obstaclesIntersectPlan(
...
+			// get the current position of the base
+			currentPosition, err := mr.kinematicBase.CurrentPosition(ctx)
+			if err != nil {
+				return state.ExecuteResponse{}, err
+			}
+
+			// Note: the value of wayPointIndex is subject to change between when this function is first entered
+			// versus when CheckPlan is actually called.
+			// We load the wayPointIndex value to ensure that all information is up to date.
+			waypointIndex := int(mr.waypointIndex.Load())
```
Comment (biotinker) [inline services/motion/builtin/move_request.go:303]: This seems like a race. If execution swaps to the next trajectory in between getting `currentInputs` and getting `waypointIndex` then these will not match.

I think we need a mutex protecting this so that `currentInputs` and `waypointIndex` can be guaranteed to match.

File: services/motion/builtin/move_request.go (hunk 4 above)
Comment (biotinker) [reply services/motion/builtin/move_request.go:140]: I don't think this actually fixed the race, but rather simply moved it elsewhere; see comment here:

https://github.com/viamrobotics/rdk/pull/3718/files#r1535927476

File: motionplan/motionPlanner.go (hunk 3 above)
Comment (nfranczak) [reply motionplan/motionPlanner.go:540]: > I am wondering if perhaps we want to instead pass in entire plans, and pass in the currently executing index, instead. Not necessarily in this PR.

I think this is a good way to solve this issue. I will incorporate this change and push it up.

**Review by biotinker — APPROVED** [review]
LGTM, nice work on this

File: services/motion/builtin/move_request.go (hunk 4 above)
Comment (nfranczak) [reply services/motion/builtin/move_request.go:140]: I will fix the race, thank you for noticing that it was not correctly handled.

**Review by raybjork — APPROVED** [review]
Nice job this looks good
